beacon-chain/sync/data_columns_sampling.go

```diff
@@ -50,205 +56,285 @@ func sortedListFromMap(m map[uint64]bool) []uint64 {
+			}
+		}
+
+		roundSummaries = append(roundSummaries, roundSummary{
+			RequestedColumns: columnsToSample,
+			MissingColumns:   missingSamples,
+		})
+
+		retrievedSampleCount := uint64(len(retrievedSamples))
+
+		if retrievedSampleCount == columnsToSampleCount {
+			// All columns were correctly sampled, this is the happy path.
+			log.WithFields(logrus.Fields{
+				"root":         fmt.Sprintf("%#x", root),
+				"roundsNeeded": round,
+			}).Debug("All columns were successfully sampled")
+			return true, roundSummaries, nil
+		}
 
-	return nil
+		if retrievedSampleCount > columnsToSampleCount {
+			// This should never happen.
+			return false, nil, errors.New("retrieved more columns than requested")
+		}
+
+		// Some columns are missing, we need to extend the sample size.
+		missingColumnsCount += columnsToSampleCount - retrievedSampleCount
+
+		firstColumnToSample = extendedSampleCount
+		oldExtendedSampleCount := extendedSampleCount
+		extendedSampleCount = peerdas.ExtendedSampleCount(sampleCount, missingColumnsCount)
+
+		log.WithFields(logrus.Fields{
+			"root":                fmt.Sprintf("%#x", root),
+			"round":               round,
+			"missingColumnsCount": missingColumnsCount,
+			"currentSampleCount":  oldExtendedSampleCount,
+			"nextSampleCount":     extendedSampleCount,
+		}).Debug("Some columns are still missing after sampling this round.")
+	}
 }
 
-func (s *Service) dataColumnSampling(ctx context.Context) {
+// DataColumnSamplingRoutine runs incremental DAS on block when received.
+func (s *Service) DataColumnSamplingRoutine(ctx context.Context) {
+	// Get the custody subnets count.
+	custodySubnetsCount := peerdas.CustodySubnetCount()
+
 	// Create a subscription to the state feed.
 	stateChannel := make(chan *feed.Event, 1)
 	stateSub := s.cfg.stateNotifier.StateFeed().Subscribe(stateChannel)
 
 	// Unsubscribe from the state feed when the function returns.
 	defer stateSub.Unsubscribe()
 
+	// Retrieve the number of columns.
+	columnsCount := params.BeaconConfig().NumberOfColumns
+
+	// Retrieve all columns we custody.
+	custodyColumns, err := peerdas.CustodyColumns(s.cfg.p2p.NodeID(), custodySubnetsCount)
+	if err != nil {
+		log.WithError(err).Error("Failed to get custody columns")
+		return
+	}
+
+	custodyColumnsCount := uint64(len(custodyColumns))
+
+	// Compute the number of columns to sample.
+	if custodyColumnsCount >= columnsCount/2 {
+		log.WithFields(logrus.Fields{
+			"custodyColumnsCount": custodyColumnsCount,
+			"columnsCount":        columnsCount,
+		}).Debug("The node custodies at least the half the data columns, no need to sample")
+		return
+	}
+
+	samplesCount := min(params.BeaconConfig().SamplesPerSlot, columnsCount/2-custodyColumnsCount)
+
+	// Compute all the columns we do NOT custody.
+	nonCustodyColums := make(map[uint64]bool, columnsCount-custodyColumnsCount)
+	for i := uint64(0); i < columnsCount; i++ {
+		if !custodyColumns[i] {
+			nonCustodyColums[i] = true
+		}
+	}
+
+	log.Warning("CCCCCC")
```

Comment:
this is not needed

Reply:
My bad. Fixed in 655f926f0e868abc5ae30fd175412acb78a1f2c7.